mesonbuild/interpreter.py

```diff
@@ -4572,6 +4572,28 @@ def build_incdir_object(self, incdir_strings, is_system=False):
 put in the include directories by default so you only need to do
 include_directories('.') if you intend to use the result in a
 different subdirectory.
+''')
+            else:
+                try:
+                    self.validate_within_subproject(a, '')
+                except InterpreterException:
+                    mlog.warning('include_directories sandbox violation!')
+                    print(f'''The project is trying to access the directory {a} which belongs to a different
+subproject. This is a problem as it hardcodes the relative paths of these two projeccts.
+This makes it impossible to compile the project in any other directory layout and also
+prevents the subproject from changing its own directory layout.
+
+Instead of poking directly at the internals the subproject should be executed and 
+it should set a variable that the caller can then use. Something like:
+
+# In subproject
+my_inc = include_directories('include')
+
+# In parent project
+dep_sp = subproject('sub')
+dep_inc = dep_sp.get_variable('my_inc')
```

Comment:
I would rather recommend using declare_dependency() and use the subproject as dependency() fallback. IMHO we should even deprecate or at least no recommend subp.get_variable(), it's also grabbing a subproject internals.

Reply:
For this case `declare_dependency()` would be better, but I think there are legitimate use cases for `subproject.get_variable()` too where it would be awkward to make a dummy dependency to export some variables/values.

So not convinced we should deprecate `subproject.get_variable()`.

Reply:
Deprecate is indeed probably too much, but I would not recommend it in this warning at least.